### PR TITLE
Hysteretic utils added to PorousFlow

### LINKED ISSUE #16078: PorousFlow: utils for hysteretic vanGenuchten functions
## Reason

The hysteretic capillary pressure and relative permeability functions are of the vanGenuchten form, but have enhancements over the functions already in `src/utils/PorousFlowVanGenuchten.C`.  To implement hysteresis, that class needs to be enhanced

## Design

Add functions to `src/utils/PorousFlowVanGenuchten.C`

## Impact

Paving the way for hysteresis in PorousFlow

### PR DESCRIPTION
**Pull request by WilkAndy**
Fixes #16078

### CODE REVIEW

**Comment by WilkAndy** [comment]
This PR needs a review, @aeslaughter .

File: modules/porous_flow/include/utils/PorousFlowVanGenuchten.h [hunk 1]
```diff
@@ -137,4 +137,135 @@ Real dRelativePermeabilityNW(Real seff, Real m);
...
+{
+  enum ExtensionStrategy
+  {
+    NONE,
+    QUADRATIC,
+    EXPONENTIAL
+  } strategy;
+  Real S;
+  Real Pc;
+  Real dPc;
```
Comment (aeslaughter) [inline modules/porous_flow/include/utils/PorousFlowVanGenuchten.h:159]: For clarity please separate the definition of the enum from the member variable.
Can/should these be const members?

```suggestion
  enum ExtensionStrategy
  {
    NONE,
    QUADRATIC,
    EXPONENTIAL
  };
  const ExtensionStrategy strategy;
  const Real S;
  const Real Pc;
  const Real dPc;
```

File: modules/porous_flow/include/utils/PorousFlowVanGenuchten.h [hunk 2]
```diff
@@ -137,4 +137,135 @@ Real dRelativePermeabilityNW(Real seff, Real m);
...
+
+/**
+ * Parameters associated with the extension of the hysteretic wetting capillary pressure function to
+ * high saturation values
+ * @ ExtensionStrategy the type of extension used
+ * @ S liquid saturation at the point of extension
+ * @ Pc capillary pressure at the point of extension
+ * @ dPc d(Pc)/dS at the point of extension
+ */
+struct HighCapillaryPressureExtension
```
Comment (aeslaughter) [inline modules/porous_flow/include/utils/PorousFlowVanGenuchten.h:176]: Same statement and question as above

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 3]
```diff
@@ -203,4 +203,221 @@ d2RelativePermeabilityNW(Real seff, Real m)
...
+Real
+capillaryPressureHys(Real sl,
+                     Real slmin,
+                     Real sgrdel,
+                     Real alpha,
+                     Real n,
+                     const LowCapillaryPressureExtension & low_ext,
+                     const HighCapillaryPressureExtension & high_ext)
+{
+  if (sl < low_ext.S) // important for initializing low_ext that this is < and not <=
```
Comment (aeslaughter) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:216]: All these return values makes this confusing. Would it be better to make this an if, else if, else function and define the output up front.

```
Real output;
if (sl < low_ext.S)
  output = ...
else if (sl >high_ext.S)
  output = ...
else
  output = ...
return output;

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 4]
```diff
@@ -203,4 +203,221 @@ d2RelativePermeabilityNW(Real seff, Real m)
...
+Real
+dcapillaryPressureHys(Real sl,
+                      Real slmin,
+                      Real sgrdel,
+                      Real alpha,
+                      Real n,
+                      const LowCapillaryPressureExtension & low_ext,
+                      const HighCapillaryPressureExtension & high_ext)
+{
+  if (sl < low_ext.S) // important for initializing low_ext that this is < and not <=
```
Comment (aeslaughter) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:264]: Same comment as above

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 5]
```diff
@@ -203,4 +203,221 @@ d2RelativePermeabilityNW(Real seff, Real m)
...
+
+Real
+saturationHys(Real pc,
+              Real slmin,
+              Real sgrdel,
+              Real alpha,
+              Real n,
+              const LowCapillaryPressureExtension & low_ext,
+              const HighCapillaryPressureExtension & high_ext)
+{
```
Comment (aeslaughter) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:315]: same comment as above

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 6]
```diff
@@ -203,4 +203,221 @@ d2RelativePermeabilityNW(Real seff, Real m)
...
+  }
+  if (pc == std::numeric_limits<Real>::max())
+    return 0.0;
+  const Real seffpow = 1.0 + std::pow(pc * alpha, n);
+  const Real seff = std::pow(seffpow, (1.0 - n) / n);
+  return (1.0 - sgrdel - slmin) * seff + slmin;
+}
+
+Real
+dsaturationHys(Real pc,
```
Comment (aeslaughter) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:365]: same

**Review by aeslaughter — COMMENTED** [review]
I have one major design question on the internals of these functions.

File: modules/porous_flow/include/utils/PorousFlowVanGenuchten.h (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/include/utils/PorousFlowVanGenuchten.h:159]: Separation is done.  I *think* the members can be const - i'll try

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 3 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:216]: done

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 4 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:264]: done

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 5 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:315]: done

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 6 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:365]: done

File: modules/porous_flow/include/utils/PorousFlowVanGenuchten.h (hunk 1 above)
Comment (aeslaughter) [reply modules/porous_flow/include/utils/PorousFlowVanGenuchten.h:159]: I don't know that they need to be. It was more of a question. Do you envision a use where you would create one of these structs and need to change values. If that doesn't make sense then make them const, if it does leave it alone.

Comment (WilkAndy) [reply modules/porous_flow/include/utils/PorousFlowVanGenuchten.h:159]: I like the idea of const, @aeslaughter , thanks.   In the current code i think it makes sense.  I'm just re-writing stuff, as with const i need some constructors.  Later on if const becomes annoying i can always remove it.

Comment (WilkAndy) [reply modules/porous_flow/include/utils/PorousFlowVanGenuchten.h]: done

File: modules/porous_flow/include/utils/PorousFlowVanGenuchten.h (hunk 2 above)
Comment (WilkAndy) [reply modules/porous_flow/include/utils/PorousFlowVanGenuchten.h:181]: done

**Review by aeslaughter — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for the super-speedy review and merge, @aeslaughter !

**Comment by aeslaughter** [comment]
No problem, thanks for all your work.

**Pull request merged**